apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs

```diff
@@ -0,0 +1,311 @@
+            await StartSessionCreating(cancellationToken).ConfigureAwait(false);
+            try
+            {
+                var sessionPoolKey = new SessionPoolKey {
+                    Client = spannerClient,
+                    Project = project,
+                    Instance = spannerInstance,
+                    Database = spannerDatabase
+                };
+                SessionPoolImpl targetPool = GlobalPool.GetOrAdd(sessionPoolKey, key => new SessionPoolImpl(key));
+
+                var session = await targetPool.AcquireSessionAsync(options, cancellationToken);
+                if (session != null)
+                {
+                    //refresh the mru list which tells us what sessions need to be trimmed from the pool when we want
+                    // to add another poolEntry.
+                    ReSortMru(targetPool);
+                    s_sessionsInUse.AddOrUpdate(session, x => sessionPoolKey, (x, y) => sessionPoolKey);
+                    LogSessionsInUse();
+                }
+                return session;
+            }
+            finally
+            {
+                EndSessionCreating();
+            }
+        }
+
+        private static void SignalAnyWaitingRequests()
+        {
+            TaskCompletionSource<int> waitingSessionRequest;
+            //if anyone is waiting, let them query for their session.
+            if (CurrentActiveSessions < MaximumActiveSessions
+                && s_waitQueue.TryDequeue(out waitingSessionRequest))
+            {
+                waitingSessionRequest.SetResult(0);
+            }
+        }
+
+        private static void LogSessionsInUse()
+        {
+            Logger.LogPerformanceCounter("Session.ActiveCount", () => s_sessionsInUse.Count);
+        }
+
+        /// <summary>
+        /// Releases a session back into the pool, possibly causing another entry to be evicted if the maximum pool size has been
+        /// reached.
+        /// </summary>
+        /// <param name="session"></param>
+        /// <param name="client"></param>
+        /// <returns></returns>
+        public static void ReleaseToPool(this SpannerClient client, Session session )
+        {
+            session.ThrowIfNull(nameof(session));
+
+            SessionPoolKey result;
+            if (s_sessionsInUse.TryRemove(session, out result))
+            {
+                LogSessionsInUse();
+                SignalAnyWaitingRequests();
+                SessionPoolImpl targetPool = GlobalPool.GetOrAdd(result,
+                                             key => new SessionPoolImpl(key));
+
+                SpannerClient evictionClient = result.Client;
+                Session evictionSession = null;
+
+                //Figure out if we want to pool this released session.
+                lock (s_priorityListSync)
+                {
+                    targetPool.ReleaseSessionToPool(client, session);
+                    if (CurrentPooledSessions > MaximumPooledSessions)
```

Comment:
That's a condition that sounds like it shouldn't happen. I thought we stopped that many sessions from being created. Or is this addressing the case where the limits are changed after creating sessions? If so, it's not clear that releasing one session should be the trigger for evicting others. Hmm.